### ISSUE TITLE
Improved accuracy in higher dimensions with alternate intercept distribution

### ISSUE THREAD

**Issue by tcstewar** [nengo/nengo_extras#34]
So, Aaron's talk reminded me about something I'd wanted to look at ages ago: in high dimensions, it might not make sense to have a uniform distribution of intercepts.  A neuron with an intercept of 0.8 in 16-dimensions will only fire for a very small proportion of points in the 16-D space (0.00196%, as it turns out). And this also means a neuron with an intercept of -0.8 will fire for 99.998% of the space.  This means that those 40% of neurons (and likely many more) are kind of useless.

So, I finally sat down an analyzed this and figured out the math of what the intercept distribution should be such that the proportion of the space for which a neuron is active is a uniform distribution.  It turns out the transformation is

```sqrt(1-betaincinv((d+1)/2, 1/2, x+1))```

(where x is the original intercept, uniformly distributed between -1 and 1, and d is the dimensionality)

I've got the derivation here: https://github.com/tcstewar/testing_notebooks/blob/master/Intercept%20Distribution%20.ipynb along with some analysis showing that this makes a pretty big improvement in decoder accuracy for higher dimensions.  Oddly enough, it does make decoding a constant function slightly worse, and it seems to make 2 and 4-dimensional linear functions a very very small amount worse.  The graphs below show the decoder error for various functions (constant (1), linear (x_i), squaring (x_i^2), and other quadratics (x_i * y_i) both with and without this system.  The green bars are with this system.

For a constant function, we actually get worse with this method.
![image](https://cloud.githubusercontent.com/assets/2380772/20647394/6c8784f2-b461-11e6-8c84-150abc431138.png)

For a linear function, we get a touch worse up until D=4, and then we get better
![image](https://cloud.githubusercontent.com/assets/2380772/20647397/890740f4-b461-11e6-81f9-a8ffa9e37dd4.png)

For a squaring function, we get an improvement everywhere
![image](https://cloud.githubusercontent.com/assets/2380772/20647398/a38d6246-b461-11e6-97a9-c9afa7ecc863.png)

And same for the other pairwise quadratics
![image](https://cloud.githubusercontent.com/assets/2380772/20647400/ae5489fc-b461-11e6-96cf-e4cb0517a7cb.png)

The implementation of this ends up being fairly compact, but it does rely on scipy:

```python
class AreaIntercepts(nengo.dists.Distribution):
    dimensions = nengo.params.NumberParam('dimensions')
    base = nengo.dists.DistributionParam('base')
    
    def __init__(self, dimensions, base=nengo.dists.Uniform(-1, 1)):
        super(AreaIntercepts, self).__init__()
        self.dimensions = dimensions
        self.base = base
        
    def __repr(self):
        return "AreaIntercepts(dimensions=%r, base=%r)" % (self.dimensions, self.base)
    
    def transform(self, x):
        sign = 1
        if x > 0:
            x = -x
            sign = -1
        return sign * np.sqrt(1-scipy.special.betaincinv((self.dimensions+1)/2.0, 0.5, x+1))
    
    def sample(self, n, d=None, rng=np.random):
        s = self.base.sample(n=n, d=d, rng=rng)
        for i in range(len(s)):
            s[i] = self.transform(s[i])
        return s
```

This feels like something that might help out SPA models a fair bit....

**Comment by Seanny123** [comment]
That's super neato, although I'm not super clear on how it's related to `spa-opt`. If we were to include it into this repository, would you want it to be conditionally applied to non-constant functions above a certain dimension or would we just trust the user to know when to use it?

**Comment by tcstewar** [comment]
> I'm not super clear on how it's related to spa-opt

I'm not sure either, but the emphasis in ```spa-opt``` was in adjusting the radius, not the intercepts.  So it might be completely independent and complementary.

> If we were to include it into this repository, would you want it to be conditionally applied to non-constant functions above a certain dimension or would we just trust the user to know when to use it?

Initially, my plan would be to have a helper function, so you could do something like:

```python
model = nengo.Network()
with model:
    ...

nengo_extras.convert_intercepts_to_area(model)
```

(I'm not sure what the right name would be).  Then people could use that, or they could manually use the AreaIntercepts() distribution.

If, it turns out that this is incredibly useful and we want to start having it automatically applied, then I think that'd have to be something in core nengo itself.  That'd be an interesting debate, since the current design for the intercepts does not have access to the Ensemble dimensionality, so we can't do it by just changing the default intercept distribution.  But I'd definitely want to hold off on any discussion like that until we've used it a lot more. That said, this seems like a lot more of a natural intercept distribution to me than ```nengo.dists.Uniform(-1,1)```.  But it's also much more complicated to explain.

(Note that this is exactly the same reasoning I argued for in spa-opt: we start by having it be a helper function, and then only worry about changing the internals of nengo if we've used it a lot and really want to make it default).

**Comment by tcstewar** [comment]
Also, I'd envision the helper function applying this transform *everywhere*.   The small drop in accuracy for constant functions and the tiny drop for low-dimensional linear functions seems to me to be made up for by the simplicity of just always doing a uniform distribution in active area.  But, again, this is something that I'd want to see how it works in practice with the helper function before making a big decision about having it happen automatically.

**Comment by arvoelke** [comment]
This sounds a lot like [CosineSimilarity](https://github.com/nengo/nengo/blob/614e7657afd1f16b296a06068f3d4673e5b575d2/nengo/dists.py#L431) and so I randomly played with the parameters until I got them to agree. Not sure why this works though (why dimension + 2)?

```
d = 3
n = 100000
dist = AreaIntercepts(d)
x1 = dist.sample(n)
x2 = nengo.dists.CosineSimilarity(dist.dimensions + 2).ppf(np.linspace(0, 1, n))

pylab.figure()
pylab.hist(x1, alpha=0.5, bins=50, label="AreaIntercepts")
pylab.hist(x2, alpha=0.5, bins=50, label="CosineSimilarity")
pylab.legend()
pylab.show()
```

![download](https://cloud.githubusercontent.com/assets/5420057/20655539/7596db16-b4f1-11e6-9331-389fb57c710e.png)

**Comment by tcstewar** [comment]
That'd make a lot of sense that they're related....  I can't quite get the proof, but they're both doing fairly similar things with ```betaincinv```.....  Hmm.....

**Comment by tcstewar** [comment]
(for those that are interested, the key thing with the ```CosineSimilarity``` thing is probably related to this bit that's right in our documentation:

```
    This can be used to calculate an intercept ``c = ppf(1 - p)`` such that
    ``dot(u, v) >= c`` with probability ``p``, for random unit vectors ``u``
    and ``v``. In other words, a neuron with intercept ``ppf(1 - p)`` will
    fire with probability ``p`` for a random unit length input.
```

**Comment by jgosmann** [comment]
I have to look into this in detail (which will probably take me a couple of days), but I'm pretty sure that there is some relation to the distributions that the spa-opt optimization is based on.

I also would like to test out what the combined effect of these optimizations is.

**Comment by tcstewar** [comment]
> I have to look into this in detail (which will probably take me a couple of days), but I'm pretty sure that there is some relation to the distributions that the spa-opt optimization is based on.
> 
> I also would like to test out what the combined effect of these optimizations is.

Cool!  I simplified the math a bit (a lot) from when I chatted with you before on it, and I'm curious what the relationship is....

**Comment by arvoelke** [comment]
All righty... so I figured out the connection(s) pretty thoroughly.

First off, it is fairly straight-forward to prove the equivalence stated in my last post using the following facts:
 - `CosineSimilarity(d + 2)` will call `SubvectorLength(d + 2, 1)` which will in turn call `SqrtBeta(n=(d + 2) - 1, m=1)`, in effect setting `n = d + 1` and `m = 1` for `SqrtBeta` where `d` is the dimensionality of `AreaIntercepts`.
 - Your `transform` method may be re-used to define the inverse CDF of your distribution as follows:
```
    def ppf(self, y):
        return self.transform(2*y - 1)  # assuming transform is vectorized
```
Note that using different base distributions in `AreaIntercepts` corresponds to different ways of [sampling from the inverse CDF](https://en.wikipedia.org/wiki/Inverse_transform_sampling).
 - `CosineSimilarity.ppf(y)` wraps `SqrtBeta.ppf(y)` as follows:
```
        x = super(CosineSimilarity, self).ppf(abs(y*2 - 1))
        return np.where(y > 0.5, x, -x)
```
 - The regularized beta funciton [has the property](https://en.wikipedia.org/wiki/Beta_function#Properties_2) `I_x(a, b) = 1 - I_{1 - x}(b, a)`.
 - Both inverse CDFs are symmetric about 1/2.

Applying these facts to half the curve shows that both inverse CDF's have the exact same formula, hence their distributions are identical. So @jgosmann it may be appropriate to think of the `AreaIntercepts` distribution as a `SqrtBeta` distribution with `n = d + 1` and `m = 1` with suitable rescaling / mirroring, or equivalently as the `CosineSimilarity`:

```
def sample_area_intercepts(dimensions, n, d=None, rng=np.random,
                           base=nengo.dists.Uniform(-1, 1)):
    return nengo.dists.CosineSimilarity(dimensions + 2).ppf(
        (base.sample(n=n, d=d, rng=rng) + 1) / 2)  # transform U[-1, 1] to U[0, 1]
```

Next... we know that `nengo.dists.CosineSimilarity(dimensions + 2)` is equivalent to `nengo.dists.UniformHypersphere(surface=True).sample(n, d + 2)[:, 0]`, i.e., the projection of a (d + 2)-dimensional unit-length vector onto one dimension (see the doc-string).

Lastly... this is also equivalent to `nengo.dists.UniformHypersphere(surface=False).sample(n, d)[:, 0]`, i.e., the projection of a d-ball onto one dimension (missing citation).

```
d = 4
n = 50000

x1 = AreaIntercepts(d).sample(n)
x2 = sample_area_intercepts(d, n)
x3 = nengo.dists.UniformHypersphere(surface=True).sample(n, d + 2)[:, 0]
x4 = nengo.dists.UniformHypersphere(surface=False).sample(n, d)[:, 0]

pylab.figure()
pylab.hist(x1, alpha=0.3, bins=50, label="AreaIntercepts")
pylab.hist(x2, alpha=0.3, bins=50, label="CosineSimilarity")
pylab.hist(x3, alpha=0.3, bins=50, label="Sphere")
pylab.hist(x4, alpha=0.3, bins=50, label="Ball")
pylab.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)
pylab.show()
```

![download 1](https://cloud.githubusercontent.com/assets/5420057/20678875/6498aa2c-b566-11e6-9bbf-2c30e05b0707.png)

In summary, all 4 of these distributions are equivalent (or 5 if you include the `SqrtBeta`). But perhaps the most obvious _in hindsight_ is the last one: we are projecting the d-dimensional ball onto one dimension. In other words this distribution is giving us a single component of the evaluation points. I think this latter perspective is the most useful.

Anyways... with all that out of the way... I like what is being done here. I'll have to give some more thought as to what makes it sometimes worse and sometimes better in your examples.

**Comment by tcstewar** [comment]
> ```x4 = nengo.dists.UniformHypersphere(surface=False).sample(n, d)[:, 0]```

Oh, *pretty*.  Yes, that's exactly right.  That's a beautiful way to express it. :) :) :)

**Comment by jgosmann** [comment]
@arvoelke was quicker than me 🏃 💨 ..... 🐌 

I just also figured out that, given the transform from intercept value to proportion of space covered, we can do a change of variables on the uniform distribution of covered space to obtain the analytic PDF of the better intercept distribution. Which turns out to be the SqrtBeta with n=d+1 and m=1 (for intercepts in [0, 1); half that for (-1, 1)).

One thing I'm still wondering about: Is there some intuition why it is SqrtBeta with n=d+1 and m=1. This is the distribution of the absolute value of a single dimension of a d+2 dimensional unit vector. How do these two extra dimensions account for the unit length (compared to `UniformHypersphere(surface=False).sample(n.d)[:, 0]`).

**Comment by tcstewar** [comment]
> ```x4 = nengo.dists.UniformHypersphere(surface=False).sample(n, d)[:, 0]```

Another very important thing about this way of expressing it is that it gives us an implementation that's not dependent on ```scipy```....

**Comment by jgosmann** [comment]
but probably less efficient as we're generating way too many samples

**Comment by jgosmann** [comment]
Another question: These calculations seem to be for encoders pointing outwards (i.e. the tuning curve doesn't cross zero). Is that right or did I miss where that is handled? But it also seems that this is not ensured in the code, but for an encoder pointing inward, the proportion of the space covered is 1-p.

**Comment by tcstewar** [comment]
In the code I posted, I just check for that case and do the one-minus

```python
        sign = 1
        if x > 0:
            x = -x
            sign = -1
```

**Comment by jgosmann** [comment]
righto
Also managed to do the analytical derivation and it just adds a factor of 1/2 to the PDF which is consistent with `CosineSimilarity`. It's beatiful when math works out. 🖼

**Comment by jgosmann** [comment]
I also took now a look on how this relates to spaopt. I only looked at d=64, subdimensions=1, and 50 neurons per dimension so far, so results might differ for other parameters.

spaopt is setting the radius which is basically just changing the evaluation points and intercepts to be uniformly distributed within the radius (actually we're not changing the intercepts, but scale the encoders, but the effect should be the same). So instead of changing the radius we could just use the `CosineSimilarity(d+2)` distribution to set the `eval_points` and `intercepts`. To me it seems that this is actually the analytically perfect solution because then the representation of each piece of the space is exactly proportional to its importance. Indeed setting intercepts and eval points in this way, improves the RMSE a lot. But it performs still slightly worse than spaopt (at least for above parameters).

At the moment I'm not completely sure why that is the case. It could be because spaopt takes the distortion error of the ensemble in account. Adding more `eval_points` or `intercepts` around a point will improve the approximation of this point (`eval_points` give it a higher weight compared to other points, `intercepts` provide more turning off/on around that point), but we get diminishing returns. If we already have lots of `eval_points` or `intercepts` around a particular point, adding another one will not improve things a lot. So it can be better to add those in areas which are less important, but were the approximation gets improved more. Thus, maybe spaopt provides a better trade-off by taking the distortion into account (the more noisy the ensemble, the less beneficial it is to have close eval_points/intercepts).

**Comment by jgosmann** [comment]
Actually, the explanation might be easier: The radius also controls the point at which the maximum firing rate is measured. With setting eval_points/intercepts, the maximum firing rate is reached for -1 and 1, whereas with spaopt it is reached at the point of the radius. This gives higher firing rates (= less noise) and more non-linear tuning curves.

**Comment by tcstewar** [comment]
> I only looked at d=64, subdimensions=1

Hmm, with subdimensions=1, the transformation we're playing with here doesn't affect anything at all (the intercepts do not change)....

**Comment by jgosmann** [comment]
Yeah, seems like when correctly adjusting for the maximum firing rates, setting eval_points/intercepts with `CosineSimilarity(d+2)` performs better than spaopt! It is also much more efficient because it is a purely analytic solution.

**Comment by jgosmann** [comment]
> Hmm, with subdimensions=1, the transformation we're playing with here doesn't affect anything at all (the intercepts do not change)....

That would be true if the intercepts were calculated on the actual ensemble dimensionality, but keep in mind that the ensemble presents only part of a larger vector and thus the intercepts have to be computed as if a d dimensional vector would be represented.

**Comment by tcstewar** [comment]
> That would be true if the intercepts were calculated on the actual ensemble dimensionality, but keep in mind that the ensemble presents only part of a larger vector and thus the intercepts have to be computed as if a d dimensional vector would be represented.

Ah, right, now I get it....  thanks!

**Comment by tcstewar** [comment]
> Yeah, seems like when correctly adjusting for the maximum firing rates, setting eval_points/intercepts with CosineSimilarity(d+2) performs better than spaopt! It is also much more efficient because it is a purely analytic solution.

!!!!!  That is very surprising to me!   very curious to look at that data....  :) :)

**Comment by jgosmann** [comment]
Which part is surprising to you? The better performance in terms of error or the higher efficiency in computation?

**Comment by jgosmann** [comment]
There is another major reason why setting eval points and intercepts is better than spaopt: spaopt requires the high-dimensional ensemble to be split up into subensembles (because the optimal spaopt radius for the full high-dimensional ensemble is 1). This eval_points/intercepts approach does not depend on this. It performs (nearly) equally well no matter what the number of subdimensions is.

**Comment by arvoelke** [comment]
> One thing I'm still wondering about: Is there some intuition why it is SqrtBeta with n=d+1 and m=1. This is the distribution of the absolute value of a single dimension of a d+2 dimensional unit vector. How do these two extra dimensions account for the unit length (compared to UniformHypersphere(surface=False).sample(n.d)[:, 0]).

Yeah that's a good question. I don't have a good intuition for this either, but we've already proven it indirectly through the chain `x4 <=> x1` (Terry's math), followed by `x1 <=> x2 <=> x3` (my Math), and therefore `x3 <=> x4`. It can also be proven directly by looking at the code for `UniformHypersphere`, and using the property: `u^(1/m) ~ Beta(m, 1)` where `u ~ U[0, 1]`. First set `X = \sum_{i=0}^{d-1} v_i^2` and `Y = \sum_{i=d-1}^{d+1} v_i^2` where `v_i ~ N(0, 1)`, to apply the chi-squared property that `X / (X + Y) ~ Beta(d/2, 1)` to get `u^(2/d) \sum_{i=0}^{d+1} v_i^2 === \sum_{i=0}^{d-1} v_i^2` (`m = d/2` from the first property). Finally `sqrt` both sides, cross-multiply, and scale by `v_0`, to get the code for `x3 == x4`.

I'd be curious to know if there's a more intuitive explanation that doesn't rely on a long `<=>` chain or properties of the beta distribution. The `d + 2` is indeed slightly mystical.

**Comment by jgosmann** [comment]
I've got at least an intuitive idea why `nengo.dists.UniformHypersphere(surface=True).sample(n, d + γ) == nengo.dists.UniformHypersphere(surface=False).sample(n, d)` with `γ > 0`. The number of dimensions can also be interpreted as degrees of freedom. When `surface=True` we reduce the degrees of freedom because we limit the vectors to the surface. As such we have to increase them for the equivalence to `surface=False`. But I have no idea why `γ = 2`. A value of 1 would seem more logical ...

**Comment by jgosmann** [comment]
I think I finally figured it out. The degrees of freedom are related to the number of dimensions, but not exactly equal (though, I'm not completely clear how degrees of freedom are actually defined, that was always a somewhat fuzzy concept in my statistics course). Assume we start with `nengo.dists.UniformHypersphere(surface=False).sample(n, d)` and set `surface=True`: This will fixes all points to the hyper-sphere (surface) and we need to increase the dimensions to get back degrees of freedom to get back the original distribution. So we can introduce another dimension (d+1), but this dimension is not independent. It has to fullfil the unit-length constraint.

<img width="66" alt="image" src="https://cloud.githubusercontent.com/assets/850157/20692419/bfcd490e-b5a5-11e6-9bd8-b0d275e0ef05.png">

This can be rearranged:

<img width="92" alt="image" src="https://cloud.githubusercontent.com/assets/850157/20692439/e4399fa4-b5a5-11e6-96aa-cfab5f050a28.png">

Which means the additional dimension is already completely determined by the other d dimensions! That means `nengo.dists.UniformHypersphere(surface=True).sample(n, d+1)` has the same  degrees of freedom as our original distribution, but it is still limit to the surface (it is a different distribution). So we need to add another degree of freedom/dimension (d+2) to get the distribution for the hyper-ball. That then gives us the required flexibility to have the values in the added dimensions not completely be determined by the other d dimensions (we're free to chose one of those dimensions and the other one will be determined by that):

<img width="131" alt="image" src="https://cloud.githubusercontent.com/assets/850157/20692525/9777773a-b5a6-11e6-9a52-37439be795b0.png">

**Comment by tcstewar** [comment]
> Which part is surprising to you? The better performance in terms of error or the higher efficiency in computation?

The better performance.  With spa-opt, we're taking into account the exact distortion for that particular sample.  So it's weird to get even more of an improvement with a method that ignores the example sample and just sticks with the analytic ideal distribution....

**Comment by jgosmann** [comment]
> With spa-opt, we're taking into account the exact distortion for that particular sample.

We're not.

**Comment by jgosmann** [comment]
To elaborate a bit on my short late-night answer: spa-opt estimates the distortion of an ensemble with the same parameters, but not necessarily the same seed/intercepts/tuning curves. Taking into account the exact distortion would require to solve for twice as many decoders as in a model without spa-opt (assuming all ensembles can be optimized), whereas the current approach allows to cache results. (I suppose, we could do with solving for the decoders once if there was a way to scale them after the fact.) I'm not sure whether it is worthwhile to take the exact seed into account. I think, we're close to the minimal error that can be achieved (but I might be wrong), The distortion measure is an approximation anyways (evaluated at the eval points, but not inbetween) and averaged over all dimensions.

**Comment by arvoelke** [comment]
> We also don't have a mathematical proof that the proposed distribution (AreaIntercepts) is the right thing to do.

Certainly. An intuition that was shared by @celiasmith, @hunse, and I was that the intercepts should be roughly correlated with the magnitude of the derivative of the desired function (projected onto the encoder). For instance, if the function is linear, then the optimal setting for ReLU is to simply be active across the entire space (intercepts at -1). But if there is a sharp change in the function, for instance if it's piecewise linear about 0, then you might want there to be intercepts at 0 as well. Linking this to the intuition from the talk I gave, we want the equivalence classes (the portions of space where the same set of neurons are active, i.e., the regions between the intercepts) to be well-approximated by quadratic monomials for the desired function.

However, I haven't tried to work this out formally within the context of this thread. I'll be revisiting this at some point but thought I'd just state the above now in case it helps.

**Comment by jgosmann** [comment]
> An intuition that was shared by @celiasmith, @hunse, and I was that the intercepts should be roughly correlated with the magnitude of the derivative of the desired function (projected onto the encoder).

The intuition makes sense, but I am not sure that I can trust that intution. For example, why does `AreaIntercepts` (which shifts intercepts towards 0) perform better with the multiplicative functions that have the lowest magnitude of change around 0? (This is true for ReLU.)

**Comment by jgosmann** [comment]
I'm looking through the issues that I assigned to me at some point and was wondering what to do with this issue. It is mostly discussion that seems to be concluded for the most part. Should we still leave this open? Is there a request for a particular feature to implement in nengo_extras in here?

**Comment by jgosmann** [comment]
I'm closing this issue because discussion of this topic should move to nengo/enhancement_proposals#10 which links to a more in depth analysis of intercept distributions and proposes a concrete action to take.
